packages/belos/src/BelosBlockCGIter.hpp

```diff
@@ -398,55 +402,50 @@ class BlockCGIter<ScalarType, MV, OP, true> :
     }
     blockSize_ = blockSize;
     initialized_ = false;
-    // Use the current blockSize_ to initialize the state storage.
-    setStateSize ();
   }
 
   template <class ScalarType, class MV, class OP>
   void BlockCGIter<ScalarType,MV,OP,true>::
-  initializeCG (CGIterationState<ScalarType,MV>& newstate)
+  initializeCG (Teuchos::RCP<CGIterationStateBase<ScalarType,MV> > newstate, Teuchos::RCP<MV> R_0)
   {
     const char prefix[] = "Belos::BlockCGIter::initialize: ";
 
     // Initialize the state storage if it isn't already.
-    if (! stateStorageInitialized_) {
-      setStateSize();
-    }
-
-    TEUCHOS_TEST_FOR_EXCEPTION
-      (! stateStorageInitialized_, std::invalid_argument,
-       prefix << "Cannot initialize state storage!");
+    Teuchos::RCP<const MV> lhsMV = lp_->getLHS();
+    Teuchos::RCP<const MV> rhsMV = lp_->getRHS();
+    Teuchos::RCP<const MV> tmp = ( (rhsMV!=Teuchos::null)? rhsMV: lhsMV );
+    TEUCHOS_ASSERT(!newstate.is_null());
+    if (!Teuchos::rcp_dynamic_cast<BlockCGIterationState<ScalarType,MV> >(newstate, true)->matches(tmp, blockSize_))
+      newstate->initialize(tmp, blockSize_);
```

Comment:
@hkthorn It's getting called using `blockSize_`. The argument `numRHS` in `BlockCGIterationState::initialize()` is a leftover from the refactor. In order to be consistent, I could just rename `numRHS` to `numVectors` in all the iteration states.

Reply:
I think that would avoid confusion in the future, if you don't mind.